### PR TITLE
Swiftify object customization methods

### PR DESCRIPTION
**Pull request by alazier**
@tgoyne @segiddins

### CODE REVIEW

File: Realm/RLMObjectBase.h [hunk 1]
```diff
@@ -38,8 +38,18 @@
...
 
-+ (NSDictionary *)defaultPropertyValues;
-+ (NSString *)primaryKey;
-+ (NSArray *)ignoredProperties;
-
 @end
+
+
+
+Class RLMObjectUtilClass(bool isSwift);
```
Comment (segiddins) [inline Realm/RLMObjectBase.h]: should be `BOOL` since this is ObjC

File: Realm/RLMObjectBase.mm [hunk 2]
```diff
@@ -289,3 +269,31 @@ - (NSUInteger)hash {
...
+
++ (NSArray *)indexedPropertiesForClass:(Class)cls {
+    return [cls indexedProperties];
+}
+
++ (NSArray *)getGenericListPropertyNames:(__unused id)obj {
+    return nil;
+}
+
+@end
```
Comment (segiddins) [inline Realm/RLMObjectBase.mm]: newline

File: Realm/Tests/Swift/SwiftObjectInterfaceTests.swift [hunk 3]
```diff
@@ -109,7 +109,7 @@ class SwiftObjectInterfaceTests: SwiftTestCase {
         realm.commitWriteTransaction()
 
         let object = SwiftDefaultObject.allObjectsInRealm(realm).firstObject() as SwiftDefaultObject
-        XCTAssertEqual(object.intCol, 2, "defaultPropertyValues should override native property default value")
+        XCTAssertEqual(object.intCol, 1, "defaultPropertyValues should override native property default value")
```
Comment (segiddins) [inline Realm/Tests/Swift/SwiftObjectInterfaceTests.swift]: update failure message

File: RealmSwift/Object.swift [hunk 4]
```diff
@@ -119,7 +119,7 @@ public class Object : RLMObjectBase, Equatable {
 
     :returns: `Array` of property names to ignore.
     */
-    public class func swiftIgnoredProperties() -> [String] { return [] }
+    public class func ignoredProperties() -> [String] { return [] }
```
Comment (segiddins) [inline RealmSwift/Object.swift]: remove the `FIXME`

File: RealmSwift/Object.swift [hunk 5]
```diff
@@ -178,6 +178,25 @@ public func == <T: Object>(lhs: T, rhs: T) -> Bool {
...
+        if let type = type as? Object.Type {
+            return type.ignoredProperties() as NSArray?
+        }
+        return nil
+    }
+    @objc private class func indexedPropertiesForClass(type: AnyClass) -> NSArray? {
+        if let type = type as? Object.Type {
+            return type.indexedProperties() as NSArray?
+        }
+        return nil
```
Comment (segiddins) [inline RealmSwift/Object.swift]: should probably assert that we never get a non-`Object` class?

File: Realm/RLMObjectBase.mm [hunk 6]
```diff
@@ -289,3 +269,31 @@ - (NSUInteger)hash {
 }
 
 @end
+
+
+
+Class RLMObjectUtilClass(bool isSwift) {
+    static Class objectUtilObjc = [RLMObjectUtil class];
+    static Class objectUtilSwift = NSClassFromString(@"RealmSwift.ObjectUtil");
+    return isSwift && objectUtilSwift ? objectUtilSwift : objectUtilObjc;
```
Comment (segiddins) [inline Realm/RLMObjectBase.mm]: Why the `&& objectUtilSwift`?

**Comment by segiddins** [comment]
Maybe some of these methods belong more naturally on `RLMObjectSchema`?

File: Realm/RLMObjectBase.mm (hunk 6 above)
Comment (alazier) [inline Realm/RLMObjectBase.mm]: objectUtilSwift doesn't exist for the old swift apis - this is how we fall back to the objc version.

Comment (segiddins) [inline Realm/RLMObjectBase.mm]: In that case, won't things explode if you have a Swift subclass of `RLMObject` and you also have `RealmSwift`?

Comment (alazier) [inline Realm/RLMObjectBase.mm]: We don't support mixing the apis. ie the old apis are not supported with the RealmSwift framework.

Comment (segiddins) [inline Realm/RLMObjectBase.mm]: Then we should find a clean way to error if the APIs are mixed

**Comment by segiddins** [comment]
This change will make it impossible to provide default values for array properties

**Pull request merged**